src/coverlet.core/Symbols/CecilSymbolHelper.cs

```diff
@@ -396,6 +412,256 @@ private bool SkipGeneratedBranchesForExceptionHandlers(MethodDefinition methodDe
+                   SkipGeneratedBranchesForAsyncIterator_DisposeCheck(instructions, instruction, currentIndex);
+        }
+
+        private bool SkipGeneratedBranchesForAsyncIterator_CheckForStateSwitch(List<Instruction> instructions, Instruction instruction, int currentIndex)
+        {
+            // The pattern we're looking for here is this one:
+            //
+            // IL_0000: ldarg.0
+            // IL_0001: ldfld int32 Test.AsyncEnumerableStateMachine/'<CreateSequenceAsync>d__0'::'<>1__state'
+            // IL_0006: stloc.0
+            // .try
+            // {
+            //     IL_0007: ldloc.0
+            //     IL_0008: ldc.i4.s -4
+            //     IL_000a: sub
+            //     IL_000b: switch (IL_0026, IL_002b, IL_002f, IL_002f, IL_002d)
+            //
+            // The "switch" instruction is the branch we want to skip.  To eliminate
+            // false positives, we'll search back for the "ldfld" of the compiler-
+            // generated "<>1__state" field, making sure it precedes it within five
+            // instructions.
+
+            if (instruction.OpCode != OpCodes.Switch)
+            {
+                return false;
+            }
+
+            int minLoadStateFieldIndex = Math.Max(0, currentIndex - 5);
+
+            for (int i = currentIndex - 1; i >= minLoadStateFieldIndex; --i)
+            {
+                if (instructions[i].OpCode == OpCodes.Ldfld &&
+                    instructions[i].Operand is FieldDefinition field &&
+                    IsCompilerGenerated(field) && field.FullName.EndsWith("__state"))
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        private bool SkipGeneratedBranchesForAsyncIterator_DisposeCheck(List<Instruction> instructions, Instruction instruction, int currentIndex)
+        {
+            // Within the compiler-generated async iterator, there are at least a
+            // couple of places where we find this pattern, in which the async
+            // iterator is checking whether it's been disposed, so it'll know to
+            // stop iterating.
+            //
+            // IL_0024: ldarg.0
+            // IL_0025: ldfld bool Test.AsyncEnumerableStateMachine/'<CreateSequenceAsync>d__0'::'<>w__disposeMode'
+            // IL_002a: brfalse.s IL_0031
+            //
+            // We'll eliminate these wherever they appear.  It's reasonable to just
+            // look for a "brfalse" or "brfalse.s" instruction, preceded immediately
+            // by "ldfld" of the compiler-generated "<>w__disposeMode" field.
+
+            if (instruction.OpCode != OpCodes.Brfalse &&
+                instruction.OpCode != OpCodes.Brfalse_S)
+            {
+                return false;
+            }
+
+            if (currentIndex >= 1 &&
+                instructions[currentIndex - 1].OpCode == OpCodes.Ldfld &&
```

Comment:
also here enforce with the load of `__disposeMode`

![image](https://user-images.githubusercontent.com/7894084/110210637-d8a5c680-7e92-11eb-92d6-c3635d989267.png)

Reply:
Indeed.  Will do.